File: graph/24.py
def is_valid(board, visited, row, col):
    return (
        row >= 0
        and row < len(board)
        and col >= 0
        and col < len(board[0])
        and visited[row][col] == 0
    )


def DFS(
    board,
    visited,
    row,
    col,
    temp,
    word,
    index,
):
    visited[row][col] = 1
    temp.append(board[row][col])

    if "".join(temp) == word:
        return True

    directions = [(0, -1), (-1, -1), (-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1)]

    for coordinate in directions:
        new_row = row + coordinate[0]
        new_col = col + coordinate[1]

        if (
            is_valid(board, visited, new_row, new_col) is True
            and word[index] == board[new_row][new_col]
        ):
            if DFS(board, visited, new_row, new_col, temp, word, index + 1) is True:
                return True

    visited[row][col] = 0
    temp.pop()

    return False


def find_words(board, dictionary):
    answer = []
    flag = False

    for word in dictionary:

        for row in range(len(board)):
            for col in range(len(board[row])):

                visited = [[0] * len(row) for row in board]
                if board[row][col] == word[0]:
                    if DFS(board, visited, row, col, [], word, 1) is True:
                        answer.append(word)
                        flag = True
                        break

            if flag is True:
                flag = False
                break

    return answer


boggie = [["G", "I", "Z"], ["U", "E", "K"], ["Q", "S", "E"]]
dictionary = ["GEEKS", "FOR", "QUIZ", "GO"]

print(find_words(boggie, dictionary))

File: graph/24.cpp
#include <bits/stdc++.h>

using namespace std;

template <typename T>
using Matrix = vector<vector<T>>;

bool isValid(const Matrix<char> &G, const Matrix<int> &visited, int row, int col) {
    return row >= 0 && row < G.size() && col >= 0 && col < G[0].size() &&
           visited[row][col] == 0;
}

bool DFS(Matrix<char> &G, Matrix<int> &visited, int row, int col, string temp,
         string word) {
    temp += G[row][col];
    visited[row][col] = 1;

    if (temp == word) return true;

    vector<pair<int, int>> directions = {{0, -1}, {-1, -1}, {-1, 0}, {-1, 1},
                                         {0, 1},  {1, 1},   {1, 0},  {1, -1}};

    for (const auto &coor : directions) {
        int newRow = row + coor.first;
        int newCol = col + coor.second;

        if (isValid(G, visited, newRow, newCol) &&
            DFS(G, visited, newRow, newCol, temp, word) == true)
            return true;
    }

    visited[row][col] = 0;
    return false;
}

vector<string> findWord(Matrix<char> &G, const vector<string> &dictionary) {
    vector<string> answer;

    for (const auto &word : dictionary) {
        Matrix<int> visited(G.size(), vector<int>(G[0].size(), 0));

        for (int row = 0; row < G.size(); ++row) {
            for (int col = 0; col < G[row].size(); ++col) {
                if (G[row][col] == word[0] && visited[row][col] == 0) {
                    if (DFS(G, visited, row, col, string(""), word))
                        answer.push_back(word);
                }
            }
        }
    }

    return answer;
}

int main() {
    Matrix<char> boggle = {{'G', 'I', 'Z'}, {'U', 'E', 'K'}, {'Q', 'S', 'E'}};
    vector<string> dictionary = {"GEEKS", "FOR", "QUIZ", "GO"};

    for (const auto &word : findWord(boggle, dictionary)) cout << word << "\n";

    return 0;
}